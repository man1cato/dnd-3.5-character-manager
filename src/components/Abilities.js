import React from 'react';

const Abilities = ({ abilities, handleChange }) => (
	<div className="grid grid--abilities">
		<h5 className="grid__col1">Ability</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col1" key={i}>{abilities[ability].name}</div>
		)}

		<h5 className="grid__col2">Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col2" key={i}>{abilities[ability].score}</div>
		)}

		<h5 className="grid__col3">Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col3" key={i}>{abilities[ability].mod}</div>
		)}

		<h5 className="grid__col4">Temp Score</h5>
		{Object.keys(abilities).map((ability, i) =>
			<input
			className="grid__col4"
			key={i}
			id={ability}
			name="abilities"
			type="text"
			value={abilities[ability].tempScore}
			onChange={handleChange}
			/>
		)}

		<h5 className="grid__col5">Temp Mod</h5>
		{Object.keys(abilities).map((ability, i) =>
			<div className="grid__col5" key={i}>{abilities[ability].tempMod}</div>
		)}
	</div>
)

export default Abilities;
