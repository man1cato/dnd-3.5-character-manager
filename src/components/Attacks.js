import React from 'react';

const Attacks = (props) => (
    <div>
        <div className="row">Base Attack Bonus: {props.bab}</div>
        <div className=" grid grid--6col">
            <h4 className="grid__left-col">Attack Type</h4>
            <h4>Base</h4>
            <div></div>
            <h4>Temp Mod</h4>
            <div></div>
            <h4>Total</h4>

            <div className="grid__left-col">Melee</div>
            <div>{props.melee.base}</div>
            <div>+</div>
            <input
                type="text"
                name="melee"
                value={props.melee.mod}
                onChange={props.onInputChange}
            />
            <div>=</div>
            <div>{props.melee.total}</div>

            <div className="grid__left-col">Ranged</div>
            <div>{props.ranged.base}</div>
            <div>+</div>
            <input
                type="text"
                name="ranged"
                value={props.ranged.mod}
                onChange={props.onInputChange}
            />
            <div>=</div>
            <div>{props.ranged.total}</div>

            <div className="grid__left-col">Grapple</div>
            <div>{props.grapple.base}</div>
            <div>+</div>
            <input
                type="text"
                name="grapple"
                value={props.grapple.mod}
                onChange={props.onInputChange}
            />
            <div>=</div>
            <div>{props.grapple.total}</div>
        </div>
    </div>
)

export default Attacks