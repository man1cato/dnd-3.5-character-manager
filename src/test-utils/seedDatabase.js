import _ from 'lodash'
import profile from './fixtures/profile'
import database from '../firebase/firebase'


export const characterOne = {
   ..._.omit(profile, ['id'])
}

export const characterTwo = {
   name: 'Korg',
   gender: 'Male',
   race: 'recoDD3gAaQXU3A56',
   jobClass: 'recbxccm8pp3fTmQw',
   alignment: 'Lawful Evil',
   level: 5,
   iconUrl: 'images/orc.svg'
}

export const characterThree = {
   name: 'Rylia',
   gender: 'Female',
   race: 'recFEpQ9zuD9Yqu7l',
   jobClass: 'recQrfCKtksnnSsaZ',
   alignment: 'Chaotic Neutral',
   level: 3,
   iconUrl: 'images/halfling.svg'
}

export const userOne = {
   uid: 'userOne',
   characterIds: ['characterOne', 'characterTwo']
}

export const userTwo = {
   uid: 'userTwo',
   characterIds: ['characterThree']
}


export default async () => {
   await database.ref('users').remove()
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[0]}`).set(characterOne)
   await database.ref(`users/${userOne.uid}/profiles/${userOne.characterIds[1]}`).set(characterTwo)
   await database.ref(`users/${userTwo.uid}/profiles/${userTwo.characterIds[0]}`).set(characterThree)
}