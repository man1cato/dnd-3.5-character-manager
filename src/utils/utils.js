import _ from 'lodash'


export const convertTextToArray = (string) => _.compact(string.split(/\n/))

export const findItemById = (storeItemArray, itemId) => storeItemArray.find((item) => item.id === itemId)

export const apiObjectToArray = (obj) => _.orderBy(Object.keys(obj).map((id) => ({
	id,
	...obj[id]
})), ['name'], ['asc'])