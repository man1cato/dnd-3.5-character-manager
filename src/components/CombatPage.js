import React from 'react'
import { connect } from 'react-redux'
import update from 'immutability-helper'

import Header from './Header'
import Footer from './Footer'
import PhysicalStats from './PhysicalStats'
import Saves from './Saves'
import Attacks from './Attacks'
import Weapons from './Weapons'
import PreparedSpells from './PreparedSpells'
import { startEditProfile } from '../actions/profile'


const setNestedProps = (prop) => {
	Object.keys(prop).forEach((key) => {
		prop[key].mod = prop[key].mod || ""
		prop[key].total = prop[key].total || prop[key].base
	})
	return prop
}

export class CombatPage extends React.Component {
	state = {
		hp: {
			base: this.props.hp.base,
			mod: this.props.hp.mod || "",
			damage: this.props.hp.damage || "",
			total: this.props.hp.total || this.props.hp.base,
		},
		initiative: {
			base: this.props.initiative.base, 
			mod: this.props.initiative.mod || "",
			total: this.props.initiative.total || this.props.initiative.base
		},
		saves: setNestedProps(this.props.saves),      
		attacks: setNestedProps(this.props.attacks)
	}  	

	handleChange = (e) => {
		const name = e.target.name
		const id = e.target.id
		let value = Number(e.target.value)
		value = (value === 0 || isNaN(value)) ? "" : value
		let total
		this.setState((prevState) => {
			if (name === "hp") {
				if (id === "damage") {
					total = this.props.hp.base + prevState.hp.mod - value
					return {
						hp: update(prevState.hp, {
						damage: { $set: value },
						total: { $set: total }
						})
					}
				} 
				total = this.props.hp.base + value - prevState.hp.damage
			} else if (name === "saves" || name === "attacks") {
				total = this.props[name][id].base + value;
				return {
					[name]: update(prevState[name], {
						[id]: {
						mod: { $set: value },
						total: { $set: total }
						}
					})
				}
			} else {
				total = this.props[name].base + value
			}
			return {
				[name]: update(prevState[name], {
					mod: { $set: value },
					total: { $set: total }          
				})
			}
		}, () => {
			this.props.startEditProfile(this.props.id, { [name]: this.state[name] })
		});
	}

	render () {
		return (
			<div className="layout">
				<Header pageTitle="Combat" />
				<div className="container container--body">

					<PhysicalStats
						hp={this.state.hp}
						ac={this.props.ac}
						initiative={this.state.initiative}
						speed={this.props.speed}
						handleChange={this.handleChange}
					/>                              

					<Saves
						saves={this.state.saves}            
						handleChange={this.handleChange}
					/>            

					<Attacks 
						attacks={this.state.attacks}
						handleChange={this.handleChange}
					/>

					<Weapons 
						weaponSet={this.props.weaponSet}
						meleeBonus={this.state.attacks.melee.total}
						rangedBonus={this.state.attacks.ranged.total}
						grappleBonus={this.state.attacks.grapple.total}
					/>

					{this.props.spellbook && 
						<PreparedSpells 
							id={this.props.id}
							spellbook={this.props.spellbook}
							spells={this.props.spells}
							startEditProfile={this.props.startEditProfile}
						/>
					}

				</div>

				<Footer />
			</div>  
		)
	}
}

const mapStateToProps = (state) => ({
  id: state.profile.id,
  hp: state.profile.hp,
  speed: state.profile.speed,
  ac: state.profile.ac,
  initiative: state.profile.initiative,
  saves: state.profile.saves,
  bab: state.profile.bab,
  attacks: state.profile.attacks,
  weaponSet: state.profile.weaponSet,
  spellbook: state.profile.spellbook,
  spells: state.spells
})

const mapDispatchToProps = (dispatch, props) => ({
  startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
});


export default connect(mapStateToProps, mapDispatchToProps)(CombatPage)