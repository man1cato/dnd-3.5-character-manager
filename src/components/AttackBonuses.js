import React, { useState, useEffect, Fragment } from 'react'
import _ from 'lodash'
import update from 'immutability-helper'

import { calcBaseMeleeBonus, calcBaseRangedBonus, calcBaseGrappleBonus, convertInputValue } from '../utils/utils'


const AttackBonuses = ({ profileAttackBonusMods, abilities, baseAttackBonuses, size, handleUpdate }) => {  
    const attackBonusBases = {
        melee: _.map(baseAttackBonuses, (bab) => calcBaseMeleeBonus(bab, abilities.str.score, size)),
        ranged: _.map(baseAttackBonuses, (bab) => calcBaseRangedBonus(bab, abilities.dex.score, size)),
        grapple: _.map(baseAttackBonuses, (bab) => calcBaseGrappleBonus(bab, abilities.str.score, size))
    }
    const [attackBonusMods, setAttackBonusMods] = useState(_.mapValues(profileAttackBonusMods, (val) => val || '')) 
    const [attackBonusTotals, setAttackBonusTotals] = useState(
        _.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val ))
    ) 
    
    useEffect(() => {
        setAttackBonusTotals(_.mapValues(attackBonusMods, (mod, key) => _.map(attackBonusBases[key], (val) => Number(mod) + val)))
        handleUpdate({ attackBonusMods })
    }, [attackBonusMods])

    return (    
        <div className="grid--6col">
            <h5 className="grid__col1">Type</h5>
            <h5>Base</h5>
            <div></div>
            <h5>Mod</h5>
            <div></div>
            <h5>Total</h5>

            {_.map(['melee', 'ranged', 'grapple'], (key) => (
                <Fragment key={key}>
                    <div className="grid__col1">{_.startCase(key)}</div>
                    <div>{attackBonusBases[key].join(' / ')}</div>
                    <div>+</div>
                    <input
                        type="number"
                        id={key}
                        value={attackBonusMods[key]}
                        onChange={(e) => {
                            setAttackBonusMods(
                                update(attackBonusMods, {
                                    [key]: { $set: convertInputValue(e.target.value) }
                                })
                            )
                        }}
                    />
                    <div>=</div>
                    <div>{attackBonusTotals[key].join(' / ')}</div>
                </Fragment>
            ))}

        </div>
    )
}


export default AttackBonuses