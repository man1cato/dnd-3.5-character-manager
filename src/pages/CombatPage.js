import React from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import PhysicalStats from '../components/PhysicalStats'
import AttackBonuses from '../components/AttackBonuses/AttackBonuses'
import Saves from '../components/Saves/Saves'
import Weapons from '../components/Weapons'
import PreparedSpells from '../components/PreparedSpells'
import { startEditProfile } from '../store/actions/profile'
import { calcAbilityMod } from '../utils/utils'


export const CombatPage = ({ profile, races, jobClasses, spells, startEditProfile }) => {
	const jobClassLevel = jobClasses[profile.jobClass].levels[profile.level]
	const race = races[profile.race]
	const handleUpdate = (updates) => {	startEditProfile(profile.id, updates) }

	return (
		<div className="container container--body">
			<div className="section">
				<h3 className="section__title">Physical Stats</h3>
				<PhysicalStats
					hp={profile.hp}
					ac={profile.ac}
					initBase={
						!!profile.abilities.dex.tempScore ? 
						calcAbilityMod(profile.abilities.dex.tempScore) : 
						calcAbilityMod(profile.abilities.dex.score)
					}
					initMod={profile.initMod}
					speed={race.speed}
					handleUpdate={handleUpdate}							
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Saving Throws</h3>
				<Saves
					saveMods={profile.saveMods}
					saveBases={jobClassLevel.saves}
					handleUpdate={handleUpdate}
				/>
			</div>

			<div className="section">       
				<h3 className="section__title">Attack Bonuses</h3>
				<AttackBonuses 
					attackBonusMods={profile.attackBonusMods}
					abilities={profile.abilities}
					baseAttackBonuses={jobClassLevel.baseAttackBonuses}
					size={race.size}
					handleUpdate={handleUpdate}
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Equipped Weapons</h3>
				<Weapons 
					weaponSet={profile.equipped.weapons}
					// meleeBonus={state.attacks.melee.total}
					// rangedBonus={state.attacks.ranged.total}
					// grappleBonus={state.attacks.grapple.total}
				/>
			</div>

			{profile.spellbook && (
				<div className="section">
					<h3 className="section__title">Prepared Spells</h3>
					<PreparedSpells 
						spellbook={profile.spellbook}
						spells={spells}
						handleUpdate={handleUpdate}
					/>
				</div>
			)}

		</div>
	)
}

const mapStateToProps = ({ profile, races, jobClasses, spells }) => ({
	profile,
	races,
	jobClasses,
	spells
})

const mapDispatchToProps = (dispatch) => ({
  	startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
})


export default connect(mapStateToProps, mapDispatchToProps)(CombatPage)