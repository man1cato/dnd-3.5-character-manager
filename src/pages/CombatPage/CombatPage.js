import React from 'react'
import { connect } from 'react-redux'
import _ from 'lodash'

import PhysicalStats from '../../components/PhysicalStats/PhysicalStats'
import AttackBonuses from '../../components/AttackBonuses/AttackBonuses'
import Saves from '../../components/Saves/Saves'
import Weapons from '../../components/Weapons/Weapons'
import PreparedSpells from '../../components/PreparedSpells/PreparedSpells'
import { startEditProfile } from '../../store/actions/profile'
import { calcAbilityMod } from '../../utils/utils'


export const CombatPage = ({ profile, races, jobClasses, spells, startEditProfile }) => {
	const jobClassLevel = jobClasses[profile.jobClass].levels[profile.level]
	const race = races[profile.race]
	const handleUpdate = updates => { 
		startEditProfile(profile.id, updates) 
	}

	return (
		<div className="container--body">
			<div className="section">
				<h3 className="section__title">Physical Stats</h3>
				<PhysicalStats
					hp={profile.hp}
					ac={profile.ac}
					initBase={
						!!profile.abilities.dex.tempScore ? 
						calcAbilityMod(profile.abilities.dex.tempScore) : 
						calcAbilityMod(profile.abilities.dex.score)
					}
					initMod={profile.initiativeMod}
					speed={race.speed}
					handleUpdate={handleUpdate}							
				/>
			</div>

			<div className="section">
				<h3 className="section__title">Saving Throws</h3>
				<Saves
					saveMods={profile.saveMods}
					saveBases={jobClassLevel.saves}
					handleUpdate={handleUpdate}
				/>
			</div>

			<div className="section">       
				<h3 className="section__title">Attack Bonuses</h3>
				<AttackBonuses 
					attackBonuses={profile.attackBonuses}
					handleUpdate={handleUpdate}
				/>
			</div>

			{profile.equipped && ( 
				<div className="section">
					<h3 className="section__title">Equipped Weapons</h3>
					<Weapons 
						equippedWeapons={profile.equipped.weapons}
						attackBonuses={profile.attackBonuses}
					/>
				</div>
			)}

			{profile.spellbook && (
				<div className="section">
					<h3 className="section__title">Prepared Spells</h3>
					<PreparedSpells 
						spellbook={profile.spellbook}
						spells={spells}
						handleUpdate={handleUpdate}
					/>
				</div>
			)}

		</div>
	)
}

const mapStateToProps = state => ({
	profile: state.profile,
	races: state.races,
	jobClasses: state.jobClasses,
	spells: state.spells
})

const mapDispatchToProps = dispatch => ({
  	startEditProfile: (id, updates) => dispatch(startEditProfile(id, updates))
})


export default connect(mapStateToProps, mapDispatchToProps)(CombatPage)