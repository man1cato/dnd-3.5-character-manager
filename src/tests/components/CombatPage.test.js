import React from 'react'
import { shallow } from 'enzyme'

import { CombatPage } from '../../components/CombatPage'
import { characterOne } from '../utils/seedDatabase'


test('should render combat page with profile data', () => {
    const wrapper = shallow(<CombatPage {...characterOne} />)
    expect(wrapper).toMatchSnapshot()
})