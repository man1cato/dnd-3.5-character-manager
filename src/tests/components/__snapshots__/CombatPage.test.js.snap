// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render combat page with profile data 1`] = `
<div
  className="layout"
>
  <Connect(Header)
    pageTitle="Combat"
  />
  <div
    className="container container--body"
  >
    <PhysicalStats
      ac={
        Object {
          "base": 14,
          "flat": 10,
          "touch": 14,
        }
      }
      handleChange={[Function]}
      hp={
        Object {
          "base": 32,
          "damage": 10,
          "mod": 6,
          "total": 32,
        }
      }
      initiative={
        Object {
          "base": 2,
          "mod": "",
          "total": 2,
        }
      }
      speed={30}
    />
    <Saves
      handleChange={[Function]}
      saves={
        Object {
          "fortitude": Object {
            "base": 1,
            "mod": "",
            "total": 1,
          },
          "reflex": Object {
            "base": 1,
            "mod": "",
            "total": 1,
          },
          "will": Object {
            "base": 4,
            "mod": "",
            "total": 4,
          },
        }
      }
    />
    <Attacks
      attacks={
        Object {
          "grapple": Object {
            "base": 4,
            "mod": "",
            "total": 4,
          },
          "melee": Object {
            "base": 4,
            "mod": "",
            "total": 4,
          },
          "ranged": Object {
            "base": 6,
            "mod": "",
            "total": 6,
          },
        }
      }
      handleChange={[Function]}
    />
    <Connect(Weapons)
      grappleBonus={4}
      meleeBonus={4}
      rangedBonus={6}
      weaponSet={
        Array [
          "recOvuth9QArkI54l",
        ]
      }
    />
    <PreparedSpells
      spellbook={
        Array [
          Object {
            "spells": Array [
              Object {
                "id": "rec8IQeYmBr3f1BMC",
                "level": 0,
                "mastered": false,
                "prepared": 1,
                "remaining": 1,
                "used": 0,
              },
              Object {
                "id": "recwpeRoCsea4vidW",
                "level": 0,
                "mastered": true,
                "prepared": 2,
                "remaining": 1,
                "used": 1,
              },
            ],
            "spellsPerDay": 4,
          },
          Object {
            "spells": Array [
              Object {
                "id": "recRhVs0ahtQZpbXx",
                "level": 1,
                "mastered": false,
                "prepared": 1,
                "remaining": 0,
                "used": 1,
              },
            ],
            "spellsPerDay": 3,
          },
        ]
      }
    />
  </div>
  <Connect(Footer) />
</div>
`;
